google-oauth-client/src/main/java/com/google/api/client/auth/openidconnect/IdTokenVerifier.java

```diff
@@ -280,5 +439,170 @@ public Builder setAcceptableTimeSkewSeconds(long acceptableTimeSkewSeconds) {
+          String publicKeyPem = (String) jwks.get(keyId);
+          keyCacheBuilder.put(keyId, buildPublicKey(publicKeyPem));
+        }
+      } else {
+        for (JsonWebKey key : jwks.keys) {
+          try {
+            keyCacheBuilder.put(key.kid, buildPublicKey(key));
+          } catch (NoSuchAlgorithmException
+              | InvalidKeySpecException
+              | InvalidParameterSpecException ignored) {
+            ignored.printStackTrace();
+          }
+        }
+      }
+
+      return keyCacheBuilder.build();
+    }
+
+    private PublicKey buildPublicKey(JsonWebKey key)
+        throws NoSuchAlgorithmException, InvalidParameterSpecException, InvalidKeySpecException {
+      if ("ES256".equals(key.alg)) {
+        return buildEs256PublicKey(key);
+      } else if ("RS256".equals((key.alg))) {
+        return buildRs256PublicKey(key);
+      } else {
+        return null;
+      }
+    }
+
+    private PublicKey buildPublicKey(String publicPem)
+        throws CertificateException, UnsupportedEncodingException {
+      return CertificateFactory.getInstance("X.509")
+          .generateCertificate(new ByteArrayInputStream(publicPem.getBytes("UTF-8")))
+          .getPublicKey();
+    }
+
+    private PublicKey buildRs256PublicKey(JsonWebKey key)
+        throws NoSuchAlgorithmException, InvalidKeySpecException {
+      com.google.common.base.Preconditions.checkArgument("RSA".equals(key.kty));
+      com.google.common.base.Preconditions.checkNotNull(key.e);
+      com.google.common.base.Preconditions.checkNotNull(key.n);
+
+      BigInteger modulus = new BigInteger(1, Base64.decodeBase64(key.n));
+      BigInteger exponent = new BigInteger(1, Base64.decodeBase64(key.e));
+
+      RSAPublicKeySpec spec = new RSAPublicKeySpec(modulus, exponent);
+      KeyFactory factory = KeyFactory.getInstance("RSA");
+      return factory.generatePublic(spec);
+    }
+
+    private PublicKey buildEs256PublicKey(JsonWebKey key)
+        throws NoSuchAlgorithmException, InvalidParameterSpecException, InvalidKeySpecException {
+      com.google.common.base.Preconditions.checkArgument("EC".equals(key.kty));
+      com.google.common.base.Preconditions.checkArgument("P-256".equals(key.crv));
+
+      BigInteger x = new BigInteger(1, Base64.decodeBase64(key.x));
+      BigInteger y = new BigInteger(1, Base64.decodeBase64(key.y));
+      ECPoint pubPoint = new ECPoint(x, y);
+      AlgorithmParameters parameters = AlgorithmParameters.getInstance("EC");
+      parameters.init(new ECGenParameterSpec("secp256r1"));
+      ECParameterSpec ecParameters = parameters.getParameterSpec(ECParameterSpec.class);
+      ECPublicKeySpec pubSpec = new ECPublicKeySpec(pubPoint, ecParameters);
+      KeyFactory kf = KeyFactory.getInstance("EC");
+      return kf.generatePublic(pubSpec);
+    }
+  }
+
+  /** Custom exception for wrapping all verification errors. */
+  public static class VerificationException extends Exception {
```

Comment:
Can you check whether this needs to be public? Given `verify` catches it without exposing it to users, package-private seems appropriate to me.

Reply:
fixed